File: Tests/Tests/STPPaymentMethodBillingDetailsTest.m
//
//  STPPaymentMethodBillingDetailsTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 3/6/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPPaymentMethodBillingDetails.h"
#import "STPFixtures.h"
#import "STPTestUtils.h"

@interface STPPaymentMethodBillingDetailsTest : XCTestCase

@end

@implementation STPPaymentMethodBillingDetailsTest

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[];
    
    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:STPTestJSONPaymentMethod][@"billing_details"] mutableCopy];
        [response removeObjectForKey:field];
        
        XCTAssertNil([STPPaymentMethodBillingDetails decodedObjectFromAPIResponse:response]);
    }
    
    XCTAssertNotNil([STPPaymentMethodBillingDetails decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:STPTestJSONPaymentMethod][@"billing_details"]]);
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:STPTestJSONPaymentMethod][@"billing_details"];
    STPPaymentMethodBillingDetails *billingDetails = [STPPaymentMethodBillingDetails decodedObjectFromAPIResponse:response];
    XCTAssertEqualObjects(billingDetails.email, @"jenny@example.com");
    XCTAssertEqualObjects(billingDetails.name, @"jenny");
    XCTAssertEqualObjects(billingDetails.phone, @"+15555555555");
    XCTAssertNotNil(billingDetails.address);
}

@end

File: Stripe/StripeiOSTests/STPPaymentMethodBillingDetailsTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPPaymentMethodBillingDetailsTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 3/6/19.
//  Copyright © 2019 Stripe, Inc. All rights reserved.
//

class STPPaymentMethodBillingDetailsTest: XCTestCase {
    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let requiredFields: [String]? = []

        for field in requiredFields ?? [] {
            var response = STPTestUtils.jsonNamed(STPTestJSONPaymentMethodCard)["billing_details"] as? [AnyHashable: Any]
            response?.removeValue(forKey: field)

            XCTAssertNil(STPPaymentMethodBillingDetails.decodedObject(fromAPIResponse: response))
        }

        XCTAssertNotNil(STPPaymentMethodBillingDetails.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed(STPTestJSONPaymentMethodCard)["billing_details"] as? [AnyHashable: Any]))
    }

    func testDecodedObjectFromAPIResponseMapping() {
        let response = STPTestUtils.jsonNamed(STPTestJSONPaymentMethodCard)["billing_details"] as? [AnyHashable: Any]
        let billingDetails = STPPaymentMethodBillingDetails.decodedObject(fromAPIResponse: response)
        XCTAssertEqual(billingDetails?.email, "jenny@example.com")
        XCTAssertEqual(billingDetails?.name, "jenny")
        XCTAssertEqual(billingDetails?.phone, "+15555555555")
        XCTAssertNotNil(billingDetails?.address)
    }
}
